components/observations/uploader/ObservationsUploader.ts

```diff
@@ -156,6 +265,65 @@ export class ObservationUploader {
     this.tryRunTaskQueue();
   }
 
+  private findTaskById(taskId: string) {
+    return this.completedTasks.find(t => t.id === taskId) || this.taskQueue.find(t => t.id === taskId);
+  }
+
+  private findTasksByParentId(taskId: string) {
+    return [...this.completedTasks.filter(t => t.parentId === taskId), ...this.taskQueue.filter(t => t.parentId === taskId)];
+  }
+
+  private setTaskStatus(taskId: string, status: TaskStatus, updateChildren = true) {
+    const task = this.findTaskById(taskId);
+
+    if (!task) {
+      this.logger.warn({taskId, status}, 'Unable to update task status - task not found');
+      return false;
+    } else if (task.status === 'working') {
+      this.logger.warn({taskId, status}, 'Unable to update task status - task is in progress');
+      return false;
+    } else if (task.status === 'error' || task.status === 'success') {
+      this.logger.warn({taskId, status}, 'Unable to update task status - task has finished');
+      return false;
+    }
+
+    task.status = status;
+
+    if (updateChildren) {
+      this.findTasksByParentId(taskId).map(t => this.setTaskStatus(t.id, status, updateChildren));
+    }
+
+    return true;
+  }
+
+  pauseUpload(observationId: string, pauseImages = true) {
+    this.setTaskStatus(observationId, 'paused', pauseImages);
+    this.tryRunTaskQueue();
+  }
+
+  resumeUpload(observationId: string, resumeImages = true) {
+    this.setTaskStatus(observationId, 'pending', resumeImages);
+    this.tryRunTaskQueue();
+  }
+
+  getUploadStatus(observationId: string): UploadStatus {
+    const task = this.findTaskById(observationId);
+    if (!task) {
+      throw new Error(`No task found with id ${observationId}`);
```

Comment:
Why do we prefer to throw here (and require the caller to try/catch) as opposed to returning an optional status and having the caller handle the missing case?

Reply:
I would always prefer to not return optionals unless I have to. And at this point I don't have a use case that motivates returning an optional value here...that may change, in which case it's easy enough to adapt.

Reply:
Isn't this line (the throw itself) where an option is useful? For the same reason that `.find()` returns one? How can we be certain that all upstream callers knows this can throw and have correctly wrapped it in `try/catch`?

Reply:
we are in philosophically different places

Reply:
As long as we can be confident that the error is handled (and doesn't crash the app at runtime) it's fine. The nice thing about the option is that `tsc` can enforce we handle it correctly.